Runtime error when subscribing to an export

- [x] Defect
 - [ ] Feature Request or Change Proposal
 
## Defects 

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included `nats-server -DV` output
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)
 
#### Versions of `nats-server` and affected client libraries used:

```
$ nats-server version
nats-server: v2.1.0
````

#### OS/Container environment:

MacOS

#### Steps or code to reproduce the issue:
 
```
# Creation of an operator
nsc add operator -n preprod

# Creation of an admin account
nsc add account --name admin

# Add an export to the account
nsc add export --account admin --name "client1" --subject "client1.>" --private

# Add user to the account
nsc add user --account admin --name admin --allow-pub ">" --allow-sub ">"

# Creation of a client account
nsc add account --name client1
CLIENT1_ACCOUNT_ID=$(nsc describe account client1 | grep 'Account ID' | cut -c 31-86)

# Add user to the account
nsc add user --account client1 --name user1 --deny-pub ">" --allow-sub "events.>"

# Securing the import
nsc generate activation -o /tmp/activation.jwt --account admin --target-account $CLIENT1_ACCOUNT_ID --subject "client1.>"

# Add an import to the account
nsc add import --account client1 --token /tmp/activation.jwt --local-subject "events.>"

# Running an account server
nats-account-server -nsc ~/.nsc/nats/preprod
```

Then I run a NATS server with the following configuration

```
$ cat ./server.conf
operator: /Users/luc/.nsc/nats/preprod/preprod.jwt
resolver: URL(http://localhost:9090/jwt/v1/accounts/)

$ nats-server -c server.conf
```

I got an error when trying to subscribe to the `events.>` subject:

```
$ nats-sub -creds ~/.nkeys/creds/preprod/client1/user1.creds "events.>"
Listening on [events.>]
Disconnected: will attempt reconnects for 10m
```

#### Expected result:
 
User connected to NATS and subscribed to the `events.>` subject.

#### Actual result:
 
````
$ nats-server -c server.conf -DV
[17544] 2019/10/16 15:31:58.066351 [INF] Starting nats-server version 2.1.0
[17544] 2019/10/16 15:31:58.066440 [DBG] Go build version go1.13.1
[17544] 2019/10/16 15:31:58.066443 [INF] Git commit [not set]
[17544] 2019/10/16 15:31:58.066446 [INF] Trusted Operators
[17544] 2019/10/16 15:31:58.066450 [INF]   System  : ""
[17544] 2019/10/16 15:31:58.066453 [INF]   Operator: "preprod"
[17544] 2019/10/16 15:31:58.066472 [INF]   Issued  : 2019-10-16 11:30:57 +0200 CEST
[17544] 2019/10/16 15:31:58.066476 [INF]   Expires : 1970-01-01 01:00:00 +0100 CET
[17544] 2019/10/16 15:31:58.066483 [WRN] Trusted Operators should utilize a System Account
[17544] 2019/10/16 15:31:58.066544 [INF] Listening for client connections on 0.0.0.0:4222
[17544] 2019/10/16 15:31:58.066547 [INF] Server id is NBCWFSGWL22DQKC4C75KPPMMQTOAMQJQE47BDFFECIQ3DAARRLK2UI5L
[17544] 2019/10/16 15:31:58.066549 [INF] Server is ready
[17544] 2019/10/16 15:31:58.066555 [DBG] Get non local IPs for "0.0.0.0"
[17544] 2019/10/16 15:31:58.066716 [DBG]  ip=192.168.5.153
[17544] 2019/10/16 15:31:58.066772 [DBG]  ip=192.168.99.1
[17544] 2019/10/16 15:31:58.066791 [DBG]  ip=192.168.64.1
[17544] 2019/10/16 15:32:02.346307 [DBG] 127.0.0.1:62459 - cid:1 - Client connection created
[17544] 2019/10/16 15:32:02.347764 [TRC] 127.0.0.1:62459 - cid:1 - <<- [CONNECT {"verbose":false,"pedantic":false,"jwt":"eyJ0eXAiOiJqd3QiLCJhbGciOiJlZDI1NTE5In0.eyJqdGkiOiJLSVUyVU83RU9BUlJOQVk2VVM3TEQySTM3SkhXUUlLUVpTUlVNTDdMVjZMNFY3UEFYNk1BIiwiaWF0IjoxNTcxMjI5MjEyLCJpc3MiOiJBRFZQQ1pTUUpCTU5LRE9HRTRUSUxFT0NaUlNHSTZDRlZMUEhXNDZYMkVSNktPT09OM01ZUUFQSiIsIm5hbWUiOiJ1c2VyMSIsInN1YiI6IlVEWVFSSklUREw2NVJYWE1OV1hUVFlKWU40RzVQSFhEUkRLUFdTRktPSlhaSlhUNjRaU0VQVTZHIiwidHlwZSI6InVzZXIiLCJuYXRzIjp7InB1YiI6eyJkZW55IjpbIlx1MDAzZSJdfSwic3ViIjp7ImFsbG93IjpbImV2ZW50cy5cdTAwM2UiXX19fQ.UgrcgQ07onJHrpgU5fKEhCJtLv4G_4oF1IjPdpBms1uMV3G-ad455G7nAnsLRnwahBy11vvsE71LZBWClQQMDQ","sig":"-AndA2UzFNoHxCPFaJtNL2e6YxrTFifQHhNoVF33D14LyN0pcc_AifYp2oY2c51LbmUCAD_mj5zE46dOW6oXBg","tls_required":false,"name":"NATS Sample Subscriber","lang":"go","version":"1.8.1","protocol":1,"echo":true}]
[17544] 2019/10/16 15:32:02.349315 [DBG] Account [ADVPCZSQJBMNKDOGE4TILEOCZRSGI6CFVLPHW46X2ER6KOOON3MYQAPJ] fetch took 1.144973ms
[17544] 2019/10/16 15:32:02.349905 [DBG] Updating account claims: ADVPCZSQJBMNKDOGE4TILEOCZRSGI6CFVLPHW46X2ER6KOOON3MYQAPJ
[17544] 2019/10/16 15:32:02.351296 [DBG] Account [ACMEIBWBQPU3P7CTIGZXP2JL6ZLPAKM72NCKFYZUIZOOL7UT2TV4XFF6] fetch took 1.372478ms
[17544] 2019/10/16 15:32:02.351504 [DBG] Updating account claims: ACMEIBWBQPU3P7CTIGZXP2JL6ZLPAKM72NCKFYZUIZOOL7UT2TV4XFF6
[17544] 2019/10/16 15:32:02.351511 [DBG] Adding stream export "client1.>" for ACMEIBWBQPU3P7CTIGZXP2JL6ZLPAKM72NCKFYZUIZOOL7UT2TV4XFF6
[17544] 2019/10/16 15:32:02.351535 [DBG] Adding stream import ACMEIBWBQPU3P7CTIGZXP2JL6ZLPAKM72NCKFYZUIZOOL7UT2TV4XFF6:"client1.>" for ADVPCZSQJBMNKDOGE4TILEOCZRSGI6CFVLPHW46X2ER6KOOON3MYQAPJ:"events.>"
[17544] 2019/10/16 15:32:02.352042 [TRC] 127.0.0.1:62459 - cid:1 - <<- [PING]
[17544] 2019/10/16 15:32:02.352047 [TRC] 127.0.0.1:62459 - cid:1 - ->> [PONG]
[17544] 2019/10/16 15:32:02.352277 [TRC] 127.0.0.1:62459 - cid:1 - <<- [SUB events.>  1]
panic: runtime error: slice bounds out of range [9:8]

goroutine 29 [running]:
github.com/nats-io/nats-server/server.(*client).addShadowSub(0xc000224000, 0xc0000c6ab0, 0xc0001c8180, 0xc000010100, 0x12, 0xc000024501, 0x12)
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/client.go:1951 +0x444
github.com/nats-io/nats-server/server.(*client).addShadowSubscriptions(0xc000224000, 0xc0001f0160, 0xc0000c6ab0, 0x0, 0xc00009fdc8)
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/client.go:1916 +0x9d6
github.com/nats-io/nats-server/server.(*client).processSub(0xc000224000, 0xc0000d2404, 0xb, 0x3fc, 0xc00027be00, 0xc000206080, 0xc000206098, 0x105fc12)
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/client.go:1827 +0x490
github.com/nats-io/nats-server/server.(*client).parse(0xc000224000, 0xc0000d2400, 0x17, 0x400, 0x17, 0x0)
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/parser.go:410 +0x1b06
github.com/nats-io/nats-server/server.(*client).readLoop(0xc000224000)
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/client.go:842 +0x2f6
github.com/nats-io/nats-server/server.(*Server).createClient.func2()
	/Users/luc/Development/src/github.com/nats-io/nats-server/server/server.go:1698 +0x2a
created by github.com/nats-io/nats-server/server.(*Server).startGoRoutine
`````

--- a/server/accounts.go
+++ b/server/accounts.go
@@ -1078,16 +1078,25 @@ func (a *Account) AddStreamImportWithClaim(account *Account, from, prefix string
 		return ErrStreamImportAuthorization
 	}
 
+	// Check prefix if it exists and make sure its a literal.
+	// Append token separator if not already present.
+	if prefix != "" {
+		// Make sure there are no wildcards here, this prefix needs to be a literal
+		// since it will be prepended to a publish subject.
+		if !subjectIsLiteral(prefix) {
+			return ErrStreamImportBadPrefix
+		}
+		if prefix[len(prefix)-1] != btsep {
+			prefix = prefix + string(btsep)
+		}
+	}
 	a.mu.Lock()
-	defer a.mu.Unlock()
 	if a.imports.streams == nil {
 		a.imports.streams = make(map[string]*streamImport)
 	}
-	if prefix != "" && prefix[len(prefix)-1] != btsep {
-		prefix = prefix + string(btsep)
-	}
 	// TODO(dlc) - collisions, etc.
 	a.imports.streams[from] = &streamImport{account, from, prefix, imClaim, false}
+	a.mu.Unlock()
 	return nil
 }
 

--- a/server/accounts_test.go
+++ b/server/accounts_test.go
@@ -698,6 +698,57 @@ func TestSimpleMapping(t *testing.T) {
 	}
 }
 
+// https://github.com/nats-io/nats-server/issues/1159
+func TestStreamImportLengthBug(t *testing.T) {
+	s, fooAcc, barAcc := simpleAccountServer(t)
+	defer s.Shutdown()
+
+	cfoo, _, _ := newClientForServer(s)
+	defer cfoo.nc.Close()
+
+	if err := cfoo.registerWithAccount(fooAcc); err != nil {
+		t.Fatalf("Error registering client with 'foo' account: %v", err)
+	}
+	cbar, _, _ := newClientForServer(s)
+	defer cbar.nc.Close()
+
+	if err := cbar.registerWithAccount(barAcc); err != nil {
+		t.Fatalf("Error registering client with 'bar' account: %v", err)
+	}
+
+	if err := cfoo.acc.AddStreamExport("client.>", nil); err != nil {
+		t.Fatalf("Error adding account export to client foo: %v", err)
+	}
+	if err := cbar.acc.AddStreamImport(fooAcc, "client.>", "events.>"); err == nil {
+		t.Fatalf("Expected an error when using a stream import prefix with a wildcard")
+	}
+
+	if err := cbar.acc.AddStreamImport(fooAcc, "client.>", "events"); err != nil {
+		t.Fatalf("Error adding account import to client bar: %v", err)
+	}
+
+	if err := cbar.parse([]byte("SUB events.> 1\r\n")); err != nil {
+		t.Fatalf("Error for client 'bar' from server: %v", err)
+	}
+
+	// Also make sure that we will get an error from a config version.
+	// JWT will be updated separately.
+	cf := createConfFile(t, []byte(`
+	accounts {
+	  foo {
+	    exports = [{stream: "client.>"}]
+	  }
+	  bar {
+	    imports = [{stream: {account: "foo", subject:"client.>"}, prefix:"events.>"}]
+	  }
+	}
+	`))
+	defer os.Remove(cf)
+	if _, err := ProcessConfigFile(cf); err == nil {
+		t.Fatalf("Expected an error with import with wildcard prefix")
+	}
+}
+
 func TestShadowSubsCleanupOnClientClose(t *testing.T) {
 	s, fooAcc, barAcc := simpleAccountServer(t)
 	defer s.Shutdown()

--- a/server/errors.go
+++ b/server/errors.go
@@ -106,6 +106,9 @@ var (
 	// ErrStreamImportAuthorization is returned when a stream import is not authorized.
 	ErrStreamImportAuthorization = errors.New("stream import not authorized")
 
+	// ErrStreamImportBadPrefix is returned when a stream import prefix contains wildcards.
+	ErrStreamImportBadPrefix = errors.New("stream import prefix can not contain wildcard tokens")
+
 	// ErrServiceImportAuthorization is returned when a service import is not authorized.
 	ErrServiceImportAuthorization = errors.New("service import not authorized")
 

